python/ray/tune/automl/search_policy.py

```diff
@@ -0,0 +1,229 @@
+from ray.tune.suggest import SearchAlgorithm
+from ray.tune.suggest.variant_generator import generate_variants
+from ray.tune.config_parser import make_parser, create_trial_from_spec
+
+
+def merge_config(path, value, config):
+    """Merge kv pair into config by expanding path separated by dot."""
+    keys = path.split('.')
+    for name in keys[:-1]:
+        if config.get(name) is None:
+            config[name] = {}
+        config = config[name]
+    # Set value corresponding to the last path name
+    config[keys[-1]] = value
+
+
+class AutoMLSearcher(SearchAlgorithm):
+    """Base class for AutoML search algorithm
+
+    It works in a round-by-round way. For each experiment round,
+    it generates a bunch of parameter config permutations, submits
+    and keeps track of them. Once all of them finish, results will
+    be fed back to the algorithm as a whole.
+    """
+
+    CONTINUE = "CONTINUE"
+    TERMINATE = "TERMINATE"
+
+    def __init__(self, search_space, reward_attr):
+        """Initialize AutoMLSearcher.
+
+        Arguments:
+            search_space (SearchSpace): The space to search.
+            reward_attr: The attribute name of the reward in the result.
+        """
+        # Pass experiment later to allow construction without this parameter
+        super(AutoMLSearcher, self).__init__()
+
+        self.search_space = search_space
+        self.reward_attr = reward_attr
+
+        self.experiment = None
+        self.summary = None
+        self.best_trial = None
+        self._is_finished = False
+        self._parser = make_parser()
+        self._undone_count = 0
+        self._running_trials = {}
+        self._completed_trials = {}
+
+        self._iteration = 0
+        self._total_trial_num = 0
+        self._start_ts = 0
+
+    def setup(self, exp, summary):
+        self.experiment = exp
+        self.summary = summary
+
+    def get_best_trial(self):
+        """Returns the Trial object with the best reward_attr"""
+        return self.best_trial
+
+    def next_trials(self):
+        if self._undone_count > 0:
+            # Last round not finished
+            return []
+
+        trials = []
+        raw_param_list, extra_arg_list = self._select()
+        if not extra_arg_list:
+            extra_arg_list = [None] * len(raw_param_list)
+
+        for param_config, extra_arg in zip(raw_param_list, extra_arg_list):
+            tag = ''
+            new_spec = copy.deepcopy(self.experiment.spec)
+            for path, value in param_config.items():
+                tag += '%s=%s-' % (path.split('.')[-1], value)
+                merge_config(path, value, new_spec['config'])
+
+            trial = create_trial_from_spec(
+                new_spec, self.experiment.name,
+                self._parser, experiment_tag=tag
+            )
+
+            # AutoML specific fields set in Trial
+            trial.results = []
+            trial.best_result = None
+            trial.param_config = param_config
```

Comment:
is this not `trial.config`?

btw I think we can later on consider making these part of Trial upstream

Reply:
Not exactly, it's the config dict in the format of flat style: key -> value, without nested dict